### PR TITLE
CONFIGURE: Bring in CFLAGS (v1.3)

### PR DESCRIPTION
**Pull request by MattBBaker**
Same change as in #1554 except for the master branch.

### CODE REVIEW

**Comment by boehms** [comment]
I don't think this is the right approach. You are not supposed to change CFLAGS. The right approach is to configure your global flags as a separate macro and then assigned them to the target i.e. in `src/ucp/Makefile.am`:
```
libucp_la_CFLAGS="-g -Wall -Werror $UCX_CFLAGS"
```

**Comment by MattBBaker** [comment]
Talked with @boehms and found a better solution. Will post soon.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1774/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3712/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1777/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3715/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:bgate:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1778/ for details.

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3716/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3722/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1786/ for details.

**Comment by MattBBaker** [comment]
@yosefe @shamisp Okay, this looks like a workable solution.

**Comment by shamisp** [comment]
Is the failure related to the change ?

**Comment by MattBBaker** [comment]
I can't find a source for the failure in the gist. It might be but I can't find the logs I need.

**Comment by shamisp** [comment]
@yosefe ?

File: config/m4/compiler.m4 [hunk 1]
```diff
@@ -156,5 +157,9 @@ CHECK_SPECIFIC_ATTRIBUTE([optimize], [NOOPTIMIZE],
 #
 # Set C++ optimization/debug flags to be the same as for C
 #
-CPPFLAGS="$CPPFLAGS -DCPU_FLAGS=\"$OPT_CFLAGS\""
-CXXFLAGS="$CFLAGS"
+BASE_CPPFLAGS="-DCPU_FLAGS=\"$OPT_CFLAGS\""
+BASE_CXXFLAGS="$CFLAGS"
```
Comment (yosefe) [inline config/m4/compiler.m4]: i think it should be BASE_CXXFLAGS="$BASE_CFLAGS"

File: src/ucs/Makefile.am [hunk 2]
```diff
@@ -1,12 +1,13 @@
...
 # Copyright (C) Mellanox Technologies Ltd. 2001-2014.  ALL RIGHTS RESERVED.
-# Copyright (C) 2014 Copyright (C) UT-Battelle, LLC. 2014. ALL RIGHTS RESERVED.
+# Copyright (C) UT-Battelle, LLC. 2014-2017. ALL RIGHTS RESERVED.
 # Copyright (C) ARM Ltd. 2016-2017.  ALL RIGHTS RESERVED.
 # See file LICENSE for terms.
 #
 
 AUTOMAKE_OPTIONS    = nostdinc # avoid collision with built-in debug.h
-AM_CPPFLAGS         = -I$(top_builddir) -I$(abs_top_srcdir)/src
+AM_CPPFLAGS         = $(BASE_CPPFLAGS) -I$(top_builddir) -I$(abs_top_srcdir)/src
```
Comment (yosefe) [inline src/ucs/Makefile.am]: shall we use libucs_la_CFLAGS and co. instead of AM_CFLAGS?

File: config/m4/compiler.m4 [hunk 3]
```diff
@@ -1,12 +1,13 @@
...
 # Copyright (C) Mellanox Technologies Ltd. 2001-2014.  ALL RIGHTS RESERVED.
+# Copyright (c) UT-Battelle, LLC. 2017. ALL RIGHTS RESERVED.
 # See file LICENSE for terms.
 #
 
 #
 # Initialize CFLAGS
 #
-CFLAGS="-g -Wall -Werror $UCX_CFLAGS"
+BASE_CFLAGS="-g -Wall -Werror $UCX_CFLAGS"
```
Comment (yosefe) [inline config/m4/compiler.m4]: UCX_CFLAGS is not defined anywhere - let's remove it
also, maybe we should define BASE_CPPFLAGS and BASE_CXXFLAGS here for better readability?

**Comment by yosefe** [comment]
the error is coming from several Linux versions, for example sles11:
```
22:54:59 Making all in src/ucm
22:54:59 make[2]: Entering directory `/scrap/jenkins/workspace/hpc-ucx-pr/label/r-vmb-sles11-sp0-x86-64/worker/0/build-test/src/ucm'
22:54:59   CC       event/libucm_la-event.lo
22:54:59   CC       malloc/libucm_la-malloc_hook.lo
22:54:59 
cc1: warnings being treated as errors
22:54:59 In file included from /scrap/jenkins/workspace/hpc-ucx-pr/label/r-vmb-sles11-sp0-x86-64/worker/0/contrib/../src/ucs/arch/cpu.h:50,
22:54:59                  from /scrap/jenkins/workspace/hpc-ucx-pr/label/r-vmb-sles11-sp0-x86-64/worker/0/contrib/../src/ucm/event/event.c:19:
22:54:59 /scrap/jenkins/workspace/hpc-ucx-pr/label/r-vmb-sles11-sp0-x86-64/worker/0/contrib/../src/ucs/arch/x86_64/cpu.h:46: error: ‘optimize’ attribute directive ignored
22:54:59 /scrap/jenkins/workspace/hpc-ucx-pr/label/r-vmb-sles11-sp0-x86-64/worker/0/contrib/../src/ucs/arch/x86_64/cpu.h:47: error: ‘optimize’ attribute directive ignored
22:54:59 make[2]: *** [event/libucm_la-event.lo] Error 1
22:54:59 make[2]: *** Waiting for unfinished jobs....
22:55:00 make[2]: Leaving directory `/scrap/jenkins/workspace/hpc-ucx-pr/label/r-vmb-sles11-sp0-x86-64/worker/0/build-test/src/ucm'
22:55:00 make[1]: *** [all-recursive] Error 1
22:55:00 make[1]: Leaving directory `/scrap/jenkins/workspace/hpc-ucx-pr/label/r-vmb-sles11-sp0-x86-64/worker/0/build-test'
22:55:00 
make: *** [all] Error 2
```

File: config/m4/compiler.m4 (hunk 3 above)
Comment (MattBBaker) [inline config/m4/compiler.m4]: The BASE_CFLAGS are added to in the script so setting them at the top would keep them from seeing the additions.

**Comment by MattBBaker** [comment]
@yosefe Okay, I think I found a fix to that. It looks like CFLAGS was being sneaked into a function that wasn't expected. This latest push should fix it.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1790/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3727/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
@yosefe ?

**Comment by shamisp** [comment]
@MattBBaker please update v1.2 backport patch and let's get it merged

**Comment by shamisp** [comment]
Fix for #1547 . v1.2 patch is one the way.

**Pull request merged**